File: 0654-Maximum Binary Tree/0654-Maximum Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        vector<TreeNode*> St;
        for (int num : nums) {
            TreeNode* node = new TreeNode(num);
            while (St.size() && St.back()->val < num) {
                node->left = St.back();
                St.pop_back();
            }
            
            if (St.size()) {
                St.back()->right = node;
            }
            St.push_back(node);
        }
        
        return St[0];            
    }
};

File: 0601-0700/0654-Maximum Binary Tree/0654-Maximum Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        St = []
        for num in nums:
            node = TreeNode(num)
            while St and St[-1].val < num:
                node.left = St.pop()
            if St:
                St[-1].right = node
            St.append(node)
        return St[0]
